File: examples/xor.rs
use neural_network::core::tensor::Tensor;
use neural_network::nn::layer::Layer;
use neural_network::nn::model::{Metrics, Model};
use neural_network::nn::neural_network::NeuralNetworkModel;
use neural_network::nn::activation::ACTIVATIONS;
use neural_network::nn::layer::INITIALIZER;

fn main() {
    let mut nn = NeuralNetworkModel::new(vec![
        Layer::new((2, 8), ACTIVATIONS::RELU, INITIALIZER::HE),
        Layer::new((8, 4), ACTIVATIONS::RELU, INITIALIZER::HE),
        Layer::new((4, 1), ACTIVATIONS::SIGMOID, INITIALIZER::XAVIER)
        ], 5.0, Some(neural_network::nn::loss::LOSS::BINARY_CROSS_ENTROPY));
    let student = Tensor::from_vec([1.0, 1.0].to_vec());
    let inputs = Tensor::from_matrix(
        &vec![
            [0.0, 0.0].to_vec(),
            [1.0, 0.0].to_vec(),
            [0.0, 1.0].to_vec(),
            [1.0, 1.0].to_vec()
        ],
    );
    let targets = Tensor::from_vec_col(vec![0.0, 1.0, 1.0, 0.0]);
    nn.fit(&inputs, &targets, 3000);
    println!("{}", nn.evaluate(&inputs, &targets));
    println!("{}", nn.predict(&student));
}

File: www/src/xor.ts
import * as wasm from 'neural_network'

let data = new Float32Array([0, 0, 1, 0,0, 1, 1, 1])
let targets = new Float32Array([0, 1, 1, 0])

let model = new wasm.NeuralNetworkJS(2, 5, new Uint32Array([2, 8, 4, 1]), [wasm.ACTIVATIONS.RELU, wasm.ACTIVATIONS.RELU, wasm.ACTIVATIONS.SIGMOID], [wasm.INITIALIZER.HE, wasm.INITIALIZER.HE, wasm.INITIALIZER.XAVIER], wasm.LOSS.BINARY_CROSS_ENTROPY)

function colorResults () {
    for (let i = 0; i < 2; i++) {
        for (let j = 0; j < 2; j++) {
            let result = model.predict(new Float32Array([i, j]))[0]
            document.getElementById(`${i}x${j}`)!.style.backgroundColor = `rgb(${150 + 100 * (1 - result)}, ${150 + 100 * result}, 150)`
        }
    }
}

function trainSteps () {
    model.fit(data, targets, 1, 0)
    colorResults()
}

document.getElementById('trainBtn')!.onclick = trainSteps
colorResults()